pallets/gear/src/queue.rs

```diff
@@ -202,6 +205,25 @@ where
             let dispatch_id = dispatch.id();
             let dispatch_reply = dispatch.reply_details().is_some();
 
+            // In case the destination is a builtin actor, process the dispatch differently.
+            if let Some(builtin_id) = T::BuiltinRouter::lookup(&program_id) {
+                let journal = match dispatch.kind() {
+                    DispatchKind::Handle => {
+                        T::BuiltinRouter::dispatch(builtin_id, dispatch.clone(), gas_limit)
+                    }
+                    _ => {
+                        // Builtin actors can only execute `handle` messages.
+                        // All other cases result in no-execution outcome.
+                        let dispatch = dispatch.into_incoming(gas_limit);
+                        let system_reservation_ctx =
+                            SystemReservationContext::from_dispatch(&dispatch);
+                        process_non_executable(dispatch, program_id, system_reservation_ctx)
+                    }
+                };
+                core_processor::handle_journal(journal, &mut ext_manager);
+                continue;
+            }
```

Comment:
I would like to suggest another implementation. The key point is that we don't need to keep builtin program ids in a storage: I've benched generation of ids and it took about `1_000_000` of gas that is much lesser than one DB read. So the idea is that gear's Config has `BuiltinRouterProvider` that creates an instance of `BuiltinRouter` which has an array with all available builtins.


```suggestion
            // In case the destination is a builtin actor, process the dispatch differently.
            let maybe_journal = match dispatch.kind() {
                DispatchKind::Handle => {
                    builtin_router.dispatch(builtin_id, dispatch.clone(), gas_limit)
                }
                _ => {
                    // Builtin actors can only execute `handle` messages.
                    // All other cases result in no-execution outcome.
                    let dispatch = dispatch.into_incoming(gas_limit);
                    let system_reservation_ctx =
                        SystemReservationContext::from_dispatch(&dispatch);

                    Some(process_non_executable(dispatch, program_id, system_reservation_ctx))
                }
            };
            
            if let Some(journal) = maybe_journal {
                core_processor::handle_journal(journal, &mut ext_manager);
                continue;
            }
```

So our `run_queue` would look like:
```rust
fn run_queue(initial_gas: Self::Gas) -> Self::Gas {
            let gas_for_builtin_router = T::BuiltinRouterProvider::required_gas();

            // Setting adjusted initial gas allowance
            GasAllowanceOf::<T>::put(initial_gas.saturating_sub(gas_for_builtin_router));
            
            let builtin_router = T::BuiltinRouterProvider::new();

            // Ext manager creation.
            // It will be processing messages execution results following its `JournalHandler` trait implementation.
            // It also will handle delayed tasks following `TasksHandler`.
            let mut ext_manager = Default::default();

            // Processing regular and delayed tasks.
            Self::process_tasks(&mut ext_manager);

            // Processing message queue.
            Self::process_queue(ext_manager, builtin_router);

            // Calculating weight burned within the block.
            initial_gas.saturating_sub(GasAllowanceOf::<T>::get())
        }
```

The next point is that `max_gas` isn't necessary: let builtins handle gas at their own. That was an issue with bls381: the maximum gas cannot be determined since it depends on the number of arguments which should be scale-decoded first.

Sorry for my vague thoughts, let's have a call to dicuss it ).

Reply:
Thanks for the input!

Regarding the `max_gas`, the reason it is there is that, unlike in usual actors, we can't track gas spent as we progress, and stop processing once the gas has run out. In the case of `builtins`, once it's started processing it will have to finish it. Sure there is no chance of running processing indefinitely (because the code of builtin actors comes from us and we are certain nothing of that kind if possible) but the `payload`, when you decode it, may still require using more gas than the user may have provided. So we should either do message processing in two steps: decode the `payload` first and find out what the exact cost will be and then do actual processing, or charge the maximum possible gas upfront and then return the unused part.
Currently, I have opted for the latter approach.